src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala

```diff
@@ -3,37 +3,43 @@ package mesosphere.marathon.core.task.update.impl.steps
 import javax.inject.Named
 
 import akka.actor.ActorRef
-import com.google.inject.Inject
+import com.google.inject.{ Inject, Provider }
 import mesosphere.marathon.MarathonSchedulerActor.ScaleApp
-import mesosphere.marathon.core.task.Task
-import mesosphere.marathon.core.task.Task.Terminated
-import mesosphere.marathon.core.task.update.TaskStatusUpdateStep
-import mesosphere.marathon.state.Timestamp
-import org.apache.mesos.Protos.TaskStatus
+import mesosphere.marathon.core.task.{ Task, TaskStateChange, TaskStateOp }
+import mesosphere.marathon.core.task.bus.MarathonTaskStatus
+import mesosphere.marathon.core.task.bus.TaskStatusObservables.TaskUpdate
+import mesosphere.marathon.core.task.update.TaskUpdateStep
 import org.slf4j.LoggerFactory
 
 import scala.concurrent.Future
 
 /**
-  * Trigger rescale of affected app if a task died.
+  * Trigger rescale of affected app if a task died or a reserved task timed out.
   */
 class ScaleAppUpdateStepImpl @Inject() (
-    @Named("schedulerActor") schedulerActor: ActorRef) extends TaskStatusUpdateStep {
+    @Named("schedulerActor") schedulerActorProvider: Provider[ActorRef]) extends TaskUpdateStep {
   private[this] val log = LoggerFactory.getLogger(getClass)
+  private[this] lazy val schedulerActor = schedulerActorProvider.get()
 
   override def name: String = "scaleApp"
 
-  override def processUpdate(timestamp: Timestamp, task: Task, status: TaskStatus): Future[_] = {
-    val taskId = task.taskId
-
-    status.getState match {
-      case Terminated(_) =>
-        // Remove from our internal list
-        log.info(s"initiating a scale check for app [${taskId.appId}] after $taskId terminated")
-        schedulerActor ! ScaleApp(taskId.appId)
+  override def processUpdate(update: TaskUpdate): Future[_] = {
+
+    val terminalOrExpungedTask: Option[Task] = {
+      (update.stateOp, update.stateChange) match {
```

Comment:
The `TaskUpdate` does not contain the oldTask option. Otherwise I would say that we need to scale if the oldTask contained a launch, and the new task option does not.

You should have covered all current cases, though.